### DISCUSSION TITLE
message function documentation description

### DISCUSSION THREAD

**Discussion by mdanb** [pyg-team/pytorch_geometric#2455] [Q&A] [answered]
The documentation mentions that the following formula for message passing is adopted in Pytorch geometric:
![Screenshot from 2021-04-23 09-11-38](https://user-images.githubusercontent.com/37456158/115876271-18f5df80-a414-11eb-9332-0c08f5a9ba7e.png)

Then, it mentions about the function `MessagePassing.message(...)` that it "Constructs messages to node i in analogy to ϕ". I looked at the implementation for `GCNConv` (specifically, `forward`) just so that I can understand the implementation under the hood better, and I noticed that `x` is linearly transformed before being passed into `propagate` (and hence `message`). However, based on my understanding, ϕ (and hence `message`) is the one "supposed" to perform this operation based on the description in the docs. I understand that it doesn't really make a difference in terms of results, but I was wondering if someone could explain the rationale behind doing the transformation in `forward` (rather than `message`). Sorry for being too pedantic! Just want to understand the design choices better, I love the library!

**Comment by rusty1s** [comment, marked as answer]
Yes, that does not make a real difference. The reason we do this here is due to the fact that it is more memory-friendly to perform the transformation node-wise instead of edge_wise, i.e.:
```python
x = x @ self.weight  # [num_nodes, out_channels]
self.propagate(edge_index, x=x)
```
is faster than doing
```python
self.propagate(edge_index, x=x)

def message(self, x_j):
    return x_j @ self.weight  # [num_edges, out_channels]
```
since usually, `|E| >> |N|`.

**Reply by mdanb** [reply]
I see, I believe I understand why it's faster, but did you actually mean to say "more memory-friendly"? It don't quite see why it would be more memory efficient, although I do see why it would be faster (since we don't perform the transformation more than once for a **given** node)

**Reply by rusty1s** [reply]
Its actually both faster and memory-friendly as the transformation on the edge-level will produce another edge-wise matrix.

**Reply by mdanb** [reply]
I think I get it. The problem that you're referring to is that x_j and the other edge_level matrix that is produced from `x_j @ self.weight` will temporarily coexist, which is not memory efficient. Is this correct, or am I missing something? Sorry if I'm a bit slow, I'm not used to thinking about memory much, until now!

**Reply by rusty1s** [reply]
That is correct :)

**Reply by mdanb** [reply]
This is somewhat related, but based on our discussion, this (which is in the docs under "Memory-Efficient Aggregations"):
![Screenshot from 2021-04-23 16-13-18](https://user-images.githubusercontent.com/37456158/115925323-f59c5600-a44e-11eb-94c4-b64f3aaca527.png)
should maybe have the `MLP` part in `forward` as `x = MLP(x)` and `message` should maybe be `x_j - x_i`? I know this is just a quick cooked up example so it's not really relevant but just thought I'd mention it (I know, I'm too picky... haha)

**Reply by rusty1s** [reply]
Actually, the MLP needs to be applied in the message function as its both dependent on source and target node representations. There is no way to out-source that as a pre-processing step since MLP(x_i) - MLP(x_j) is different from MLP(x_j - x_i), with the latter being more powerful as it can learn different messages dependent on the source node.

**Reply by rusty1s** [reply]
This holds only true for non-linear transformations though, not for linear ones.
